flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/nlp/Word2Vec.scala

```diff
@@ -0,0 +1,243 @@
+    new FitOperation[Word2Vec, T] {
+      override def fit(
+        instance: Word2Vec,
+        fitParameters: ParameterMap,
+        input: DataSet[T])
+      : Unit = {
+        val resultingParameters = instance.parameters ++ fitParameters
+        
+        val skipGrams = input
+          .flatMap(x =>
+            x.zipWithIndex
+              .map(z => {
+                val window = (scala.math.random * 100 % resultingParameters(WindowSize)).toInt
+                Context[String](
+                  z._1, x.slice(z._2 - window, z._2) ++ x.slice(z._2 +1, z._2 + window))
+              }))
+
+        val weights = new ContextEmbedder[String]
+          .setIterations(resultingParameters(Iterations))
+          .setTargetCount(resultingParameters(TargetCount))
+          .setVectorSize(resultingParameters(VectorSize))
+          .setLearningRate(resultingParameters(LearningRate))
+          .setBatchSize(resultingParameters(BatchSize))
+          .setSeed(resultingParameters(Seed))
+          .createInitialWeightsDS(instance.wordVectors, skipGrams)
+
+        instance.wordVectors = Some(weights)
+      }
+    }
+  }
+
+  /** [[TransformDataSetOperation]] for words to vectors
+    * form skipgrams from the input dataset and learn vectors against
+    * the vocabulary constructed during the fit operation
+    * returns a dataset of distinct words and their learned representations
+    *
+    * @tparam T subtype of Iterable[String]
+    * @return
+    */
+  implicit def words2Vecs[T <: Iterable[String]] = {
+    new TransformDataSetOperation[Word2Vec, T, (String, Vector[Double])] {
+      override def transformDataSet(instance: Word2Vec,
+                                    transformParameters: ParameterMap,
+                                    input: DataSet[T]): DataSet[(String, Vector[Double])] = {
+        val resultingParameters = instance.parameters ++ transformParameters
+        
+        instance.wordVectors match {
+          case Some(vectors) =>
+            val skipGrams = input
+              .flatMap(x =>
+                x.zipWithIndex
+                  .map(z => {
+                    val window = (scala.math.random * 100 % resultingParameters(WindowSize)).toInt
+                    Context[String](
+                      z._1, x.slice(z._2 - window, z._2) ++ x.slice(z._2 + 1, z._2 + window))
+                  }))
+
+            val learnedVectors = new ContextEmbedder[String]
+              .setIterations(resultingParameters(Iterations))
+              .setTargetCount(resultingParameters(TargetCount))
+              .setVectorSize(resultingParameters(VectorSize))
+              .setLearningRate(resultingParameters(LearningRate))
+              .setBatchSize(resultingParameters(BatchSize))
+              .setSeed(resultingParameters(Seed))
+              .optimize(skipGrams, instance.wordVectors)
+
+            learnedVectors
+              .flatMap(_.fetchVectors)
+          case None =>
```

Comment:
Transformation could be performed multiple times over the same model, so do you think that's ok to throw everytime exception on incoming word set for encoding only because once the model was trained incorrectly? May be we should consider some trivial default value instead of performing of heavy Exception processing?